pkg/services/ngalert/state/manager.go

```diff
@@ -99,11 +110,71 @@ func (st *Manager) Run(ctx context.Context) error {
 		case <-ctx.Done():
 			st.log.Debug("Stopping")
 			ticker.Stop()
+			if st.saveStateAsync {
+				st.stateRunnerShutdown <- struct{}{}
+				st.stateRunnerWG.Wait()
+			}
 			return ctx.Err()
 		}
 	}
 }
 
+func (st *Manager) startSync(ctx context.Context) {
+	st.stateRunnerWG.Add(1)
+	go func(ctx context.Context) {
+		ticker := st.clock.Ticker(time.Second * 30)
+	infLoop:
+		for {
+			select {
+			case <-ticker.C:
+				if err := st.fullSync(ctx); err != nil {
+					st.log.Error("Failed to do a full state sync to database", "err", err)
+				}
+			case <-st.stateRunnerShutdown:
+				st.log.Info("Stopping state sync...")
+				shutdownCtx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+				if err := st.fullSync(shutdownCtx); err != nil {
+					st.log.Error("Failed to do a full state sync to database", "err", err)
+				}
+				cancel()
+				ticker.Stop()
+				break infLoop
+			}
+		}
+		st.log.Info("State sync shut down")
+		st.stateRunnerWG.Done()
+	}(ctx)
+}
+
+func (st *Manager) fullSync(ctx context.Context) error {
+	startTime := time.Now()
+	st.log.Info("Full state sync start")
+	var instances []ngModels.AlertInstance
+	for _, state := range st.GetAll(allOrgs) {
```

Comment:
I'm not 100% confident this is thread safe. This goroutine holds a reference to a slice of pointer to state (all states in the state cache), and the states referenced by these pointers can also be updated at the same time in the scheduler goroutine following. I think you need to make a deep-copy of all states to avoid unsynchronised concurrent access which would panic the runtime?

Reply:
Mh, we also do it like this in the Prometheus API. So we might want to address it there in a further PR. I changed it for now in https://github.com/grafana/grafana/pull/74998/commits/18df441b74f250d46a45bf36b19ab863874a1782, which also makes the code much cleaner imho.

Reply:
I just built Grafana with the race detector and can confirm this race condition also exists in the Prometheus API 😢 

```
==================
WARNING: DATA RACE
Write at 0x00c0096f7e88 by goroutine 6527:
  github.com/grafana/grafana/pkg/services/ngalert/state.(*State).SetNormal()
      /Users/grobinson/go/src/github.com/grafana/grafana/pkg/services/ngalert/state/state.go:126 +0x424
  github.com/grafana/grafana/pkg/services/ngalert/state.resultNormal()
      /Users/grobinson/go/src/github.com/grafana/grafana/pkg/services/ngalert/state/state.go:205 +0x408
  github.com/grafana/grafana/pkg/services/ngalert/state.(*Manager).setNextState()
...

Previous read at 0x00c0096f7e88 by goroutine 32878:
  github.com/grafana/grafana/pkg/services/ngalert/api.PrometheusSrv.toRuleGroup()
      /Users/grobinson/go/src/github.com/grafana/grafana/pkg/services/ngalert/api/api_prometheus.go:362 +0xcd0
  github.com/grafana/grafana/pkg/services/ngalert/api.PrometheusSrv.RouteGetRuleStatuses()
```